### PR TITLE
Usart driver

### LINKED ISSUE #58: USART Drivers
# ISSUE: Current USART Driver is not thread or interrupt safe
- Implement Thread safe USART Driver
- Fix `printf` to use thread safe usart
- Implement RTOS print debugging (print: task, dbgmessage)

### PR DESCRIPTION
**Pull request by Jchisholm204**

### CODE REVIEW

File: Q24ECU/core/src/interfaces/interface_uart.c [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include "task.h"
+#include "taskHandlers.h"
+#include "nvicConfig.h"
+
+
+uart_t debug;
+
+void uart_setup(){
+    // Enable the debug usart and setup its IQR handler
+    uart_send_init(&debug, UART_DEBUG, 250000);
```
Comment (eparly) [inline Q24ECU/core/src/interfaces/interface_uart.c]: Will we ever use anything other than the `debug` USART? If so should this function be changed to accept any USART in the params of the function?

File: Q24ECU/core/config/nvicConfig.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ *      NVIC_SetPriority(USART2_IRQn, NVIC_Priority_MAX+5);
+ *      NVIC_SetPriority(USART2_IRQn, NVIC_Priority_MIN-10);
+ * 
+ */
+
+#pragma once
+
+// Highest Priority that can be given to an NVIC Interrupt
+#define NVIC_Priority_MAX 21
+// Highest Priority that can be given to an NVIC Interrupt
```
Comment (JoshWAllen) [inline Q24ECU/core/config/nvicConfig.h]:
```suggestion
// Lowest Priority that can be given to an NVIC Interrupt
```

File: Q24ECU/core/include/interfaces/interface_uart.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * @brief Initialize a USART device and its Semaphore
+ * 
+ * @param handle The USART_t typedef handle
+ * @param interface the USARTx define from CMSIS headers
+ * @param baud USART Baud Rate
+ */
+static inline void uart_send_init(uart_t *handle, USART_TypeDef *interface, unsigned long baud){
+    handle->handle = xSemaphoreCreateBinary();
```
Comment (JoshWAllen) [inline Q24ECU/core/include/interfaces/interface_uart.h]: having 'handle' as the semaphorehandle property and the parameter name is slightly confusing. Perhaps renaming one or both to be more specific would help

File: Q24ECU/core/src/interfaces/interface_uart.c (hunk 1 above)
Comment (Jchisholm204) [reply Q24ECU/core/src/interfaces/interface_uart.c]: Yes, I just added the debug port as an example for how USART should be implemented. I will change to make the code more clear

File: Q24ECU/core/config/nvicConfig.h (hunk 2 above)
Comment (Jchisholm204) [reply Q24ECU/core/config/nvicConfig.h]: thanks

File: Q24ECU/core/include/interfaces/interface_uart.h [hunk 4]
```diff
@@ -13,6 +13,10 @@
 #include "semphr.h"
 #include "stream_buffer.h"
 
+#define UART_WRITE_OK 0
+#define UART_ERR_UNDEF 1
+#define UART_ERR_ACC 2
```
Comment (JoshWAllen) [inline Q24ECU/core/include/interfaces/interface_uart.h:18]: Nit: Could be replaced with an enum - not mandatory, more of a style preference. The biggest functional difference is scoping

**Comment by JoshWAllen** [comment]
Looks good, glad to see some solid documentation and comments. I don't want to block this going in for too long - I should be good to approve after all threads are resolved.

**Comment by Jchisholm204** [comment]
Merging to move on with CAN bus drivers

**Pull request merged**